### PR TITLE
Add kraken2 data manager

### PR DESCRIPTION
**Pull request by dfornika**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection
* [ ] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

### CODE REVIEW

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.xml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    <requirements>
+        <requirement type="package" version="2.0.7_beta">kraken2</requirement>
+    </requirements>
+    <version_command>kraken2 -version | head -n 1 | awk '{print $NF}'</version_command>
+    <command detect_errors="exit_code">
+    <![CDATA[
+        python '$__tool_directory__/kraken2_build_custom.py'
+          '${out_file}'
+          --threads \${GALAXY_SLOTS:-1}
+          --db ${database_name}
```
Comment (wm75) [inline data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.xml]:
```suggestion
          --db '${database_name}'
```

This needs quoting to be safe.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        now + ")"
+    ])
+
+    database_path = database_value
+
+    args = [
+        'https://ccb.jhu.edu/software/kraken2/dl/minikraken2_' + minikraken2_version + '_8GB.tgz'
+    ]
+
+    run(['wget'] + args, target_directory)
```
Comment (wm75) [inline data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py]: Why download this to a permanent place? Can't you just put it into the temporary job working directory? IIUC, you only want to keep the unpacked data, right?

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    args = [
+        '-p',
+        database_path,
+    ]
+
+    run(['mkdir'] + args, target_directory)
```
Comment (wm75) [inline data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py]: Why not do this with `os.mkdir()` like everywhere else? Is `-p` even necessary (haven't you created the parent directory at this point)?

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+
+def _add_data_table_entry(data_manager_dict, data_table_entry, data_table_name=DATA_TABLE_NAME):
+    data_manager_dict['data_tables'] = data_manager_dict.get( 'data_tables', {} )
+    data_manager_dict['data_tables'][data_table_name] = data_manager_dict['data_tables'].get( data_table_name, [] )
+    data_manager_dict['data_tables'][data_table_name].append( data_table_entry )
+    return data_manager_dict
+
+
+def main():
```
Comment (wm75) [inline data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py]: My comments on the `build_custom` main apply here, too.

**Comment by dfornika** [comment]
Thanks for the suggestions @wm75. Will implement them ASAP. If others can do it before me, they're welcome to send a PR to my `kraken2-data-manager` branch.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+import json
+import os
+import subprocess
+import sys
+
+
+DATA_TABLE_NAME = "kraken2_databases"
+
+
+def run(args, cwd):
```
[the remark this answers is not in the document]
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py]: Thanks, I wasn't aware of that function. I've replaced all instances of my `run()` function with `subprocess.check_call()`

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    # build the index
+    kraken2_build(
+        data_manager_dict,
+        kraken2_args,
+        params,
+        target_directory
+    )
+
+    # save info to json file
+    open(args.params, 'wb').write(json.dumps(data_manager_dict))
```
[the remark this answers is not in the document]
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py]: Thanks, I've switched to text mode for each data manager.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+        else:
+            raise
+
+    data_manager_dict = {}
+
+    # build the index
+    kraken2_build(
+        data_manager_dict,
+        kraken2_args,
+        params,
```
[the remark this answers is not in the document]
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py]: True, I've removed it.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.xml (hunk 1 above)
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.xml]: Thanks. Done.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py (hunk 3 above)
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py]: Thanks, I don't recall why I ran a subprocess here. I've switched to `os.mkdir()` as suggested.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py (hunk 2 above)
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py]: Yes, I only want to keep the unpacked data. I may need some help with this. I don't have a clear idea of how to download to the temporary job working directory or how to move the unpacked data to the appropriate directory. I've made an attempt in [this commit](https://github.com/galaxyproject/tools-iuc/pull/2340/commits/d288277f09ecb5897f17b3e457b9910b102edbc2) but I'm not sure that it's correct.

**Comment by dfornika** [comment]
I've started merging these data managers into a single data manager. My plan is to add a new data manager tool xml file and python script called `kraken2_build_database` and pull functionality from the existing `kraken2_build_standard`, `kraken2_build_custom`, `kraken2_build_minikraken` and `kraken2_build_special` into that new tool. I'll hold off on removing those four existing tools until I've done a bit more testing, but I do plan on doing that.

**Comment by dfornika** [comment]
As described above, I've added a new data manager `kraken2_build_database` that currently includes two modes: `minikraken` and `standard`. The `standard` mode is currently broken due to a change in NCBI download URLs that has been fixed in `kraken2-2.0.8_beta`, which isn't currently available in bioconda.

**Comment by dfornika** [comment]
Ok I've merged the four separate data managers into one. I've tested that they all at least **start** to build a kraken database. I've been testing in a docker container so it's a bit underpowered to actually finish building some of these databases.

I think there are probably some opportunities to remove some redundancy in `kraken2_build_database.py` and there is still some testing & design decisions to be done.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+def _add_data_table_entry(data_manager_dict, data_table_name, data_table_entry):
+    data_manager_dict['data_tables'] = data_manager_dict.get( 'data_tables', {} )
+    data_manager_dict['data_tables'][ data_table_name ] = data_manager_dict['data_tables'].get( data_table_name, [] )
+    data_manager_dict['data_tables'][ data_table_name ].append( data_table_entry )
+    return data_manager_dict
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument('params')
```
[the remark this answers is not in the document]
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py]: I'm a bit confused about this suggestion. The galaxy documentation on data managers suggest reading the json file.

https://galaxyproject.org/admin/tools/data-managers/how-to/define/

It's also the way that other data managers in this repo seem to find the `target_directory`:

https://github.com/galaxyproject/tools-iuc/blob/2f544e337886709995a93d53f394919ce8e4673a/data_managers/data_manager_build_kraken_database/data_manager/make_json.py#L14

https://github.com/galaxyproject/tools-iuc/blob/2f544e337886709995a93d53f394919ce8e4673a/data_managers/data_manager_fetch_busco/data_manager/data_manager.py#L54

...though I haven't checked all of them.

**Comment by dfornika** [comment]
I've found a more powerful system to test this on. Seems to be working!

**Comment by bgruening** [comment]
@dfornika is this still WIP? Do you want to include @mvdbeek new data-manager tests?

Thanks!

**Comment by dfornika** [comment]
Yes @bgruening , I'd still call it a work-in-progress for now. I am interested in adding some tests if possible but haven't looked at how they work. I've confirmed that the minikraken and at least one of the 'special' databases (greengenes) build correctly. I set up a 'standard' database build near the end of day on Friday but I'm not able to access my testing server until Monday to check if it completed.

**Comment by dfornika** [comment]
I've taken a quick look at the tests on the `data_manager_example_blastdb_ncbi_update_blastdb` tool. I'm not sure that I'll be able to make the tests work with this data manager unless I make some changes to what is saved to the Tool Data Table. I'm currently including a timestamp in the Tool Data Table, I thought that might be useful provenance info. I'm open to suggestions on how that could be re-designed if necessary.

**Comment by mvdbeek** [comment]
I'd say https://github.com/galaxyproject/tools-iuc/tree/master/data_managers/data_manager_bowtie2_index_builder is probably the bet template for writing data manager tests. You can use all the tricks that you can use for tool output checks, for instance you can use [assert_contents](https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output-assert-contents) to make sure that the json output matches what you would expect. If you're having trouble just ping me, I can help with the test.

**Comment by dfornika** [comment]
I noticed that there's a new feature in `v2.0.8_beta` to skip downloading a large portion of the taxonomy data during database construction (the `--skip-maps` option). This makes at least one test much more manageable. I've added the test.

When testing in planemo I'm getting an error:

```
2019-05-23 12:07:58,027 ERROR [galaxy.jobs.runners.local] Job wrapper finish method failed
Traceback (most recent call last):
  File "/tmp/tmpv3vizz7u/galaxy-dev/lib/galaxy/jobs/runners/local.py", line 152, in queue_job
    self._finish_or_resubmit_job(job_state, stdout, stderr, exit_code)
  File "/tmp/tmpv3vizz7u/galaxy-dev/lib/galaxy/jobs/runners/__init__.py", line 459, in _finish_or_resubmit_job
    job_state.job_wrapper.finish(stdout, stderr, exit_code, check_output_detected_state=check_output_detected_state)
  File "/tmp/tmpv3vizz7u/galaxy-dev/lib/galaxy/jobs/__init__.py", line 1464, in finish
    self.tool.exec_after_process(self.app, inp_data, out_data, param_dict, job=job)
  File "/tmp/tmpv3vizz7u/galaxy-dev/lib/galaxy/tools/__init__.py", line 2317, in exec_after_process
    assert data_manager is not None, "Invalid data manager (%s) requested. It may have been removed before the job completed." % (data_manager_id)
AssertionError: Invalid data manager (kraken2_build_database) requested. It may have been removed before the job completed.
FAIL
```
...but otherwise it seems to be running correctly.

**Comment by mvdbeek** [comment]
If you want to test this locally you need to target the actual tool file, `planemo test data_manager/kraken2_build_database.xml` should work

**Comment by dfornika** [comment]
Thanks @mvdbeek, that worked for me. I'm seeing this both locally and in the TravisCI log:

```
History item  different than expected, difference (using diff):
( /tmp/tmp1PTVg_kraken2_custom_data_manager.json v. /tmp/tmp_Q3mjikraken2_custom_data_manager.json )
--- local_file
+++ history_data
@@ -1 +1 @@
-{"data_tables": {"kraken2_databases": [{"path": "database", "name": "database", "value": "database"}]}}
+{"data_tables": {"kraken2_databases": [{"path": "database", "name": "database", "value": "database"}]}}
```
I don't see any difference. The file in the `test-data` dir is UTF-8 encoded. I'm not sure what else might be different?

**Comment by mvdbeek** [comment]
data managers don't write out a newline. You can strip it with `head -c -1 kraken2_custom_data_manager.json`

**Comment by dfornika** [comment]
One last thing I'd like to check before this is merged: Is there a convention for setting the version numbers on these tools? I'm not sure that I've done that consistently here. I'll take a look at some other tools in this repo.

**Comment by bgruening** [comment]
@dfornika no convention so far specifically for DMs. But following the tool convention, same version as the underlying tool version, is probably a good idea.

:+1: from my site.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        if exc.errno == errno.EEXIST and os.path.isdir( target_directory ):
+            pass
+        else:
+            raise
+
+    data_manager_dict = {}
+
+    # build the index
+    kraken2_build(
+        data_manager_dict,
```
[the remark this answers is not in the document]
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py]: Thanks, I've made this change.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_custom.py]: I've decided to keep this consistent with other data managers in the repo.

File: data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py (hunk 2 above)
Comment (dfornika) [reply data_managers/data_manager_build_kraken2_database/data_manager/kraken2_build_minikraken.py]: Thanks, I've used the job working directory as suggested.

**Comment by dfornika** [comment]
I don't have any other plans to change this data manager now. If anyone has suggestions for further changes then please let me know here. Otherwise I'll request that this be merged please.

Thanks @bgruening @mvdbeek @wm75 for your help & guidance.

**Comment by bgruening** [comment]
Thanks @dfornika!

**Pull request merged**